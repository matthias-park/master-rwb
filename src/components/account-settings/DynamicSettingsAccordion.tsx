import React, { useCallback, useContext, useMemo } from 'react';
import { SettingsForm } from '../../types/api/user/ProfileSettings';
import { useForm } from 'react-hook-form';
import clsx from 'clsx';
import Accordion from 'react-bootstrap/Accordion';
import AccordionContext from 'react-bootstrap/AccordionContext';
import Form from 'react-bootstrap/Form';
import Card from 'react-bootstrap/Card';
import { useAccordionToggle } from 'react-bootstrap/AccordionToggle';
import LoadingButton from '../LoadingButton';

interface SettingProps {
  form: SettingsForm;
  onSubmit: (
    url: string,
    body: { [key: string]: string | Blob },
    formBody?: boolean,
    updateUser?: boolean,
  ) => void;
}

const FormsWithUpdateUser = [
  'disable_player',
  'disable_player_time_out',
  'permanent_disable',
];

const DynamicSettingsAccordion = ({ form, onSubmit }: SettingProps) => {
  const currentEventKey = useContext(AccordionContext);
  const accordionOnClick = useAccordionToggle(form.id);
  const { register, handleSubmit, watch, formState } = useForm({
    defaultValues: form.fields
      .filter(field => field.default ?? false)
      .reduce((obj, value) => {
        obj[value.id] = value.default;
        return obj;
      }, {}),
  });
  const watchAllFields = watch();
  const visibilityOverrideFields = useMemo(
    () =>
      form.fields
        .filter(
          field =>
            field.type === 'select' &&
            field.values?.some(value => value.additional_fields),
        )
        .reduce((obj, field) => {
          for (const value of field.values || []) {
            if (value.additional_fields) {
              for (const additionalField of value.additional_fields) {
                if (!obj[additionalField]) obj[additionalField] = {};
                if (!obj[additionalField][field.id])
                  obj[additionalField][field.id] = [];

                obj[additionalField][field.id].push(value.id.toString());
              }
            }
          }

          return obj;
        }, {}),
    [form.fields],
  );
  const updateSettingsSubmit = useCallback(
    data =>
      onSubmit(form.action, data, false, FormsWithUpdateUser.includes(form.id)),
    [],
  );

  return (
    <Card className="settings-card">
      <Card.Header
        className={clsx(
          'settings-card__header',
          currentEventKey === form.id && 'active',
        )}
      >
        <Accordion.Toggle
          data-testid="accordion-toggle"
          as="a"
          eventKey={form.id}
          className="text-dark"
        >
          {form.title}
        </Accordion.Toggle>
      </Card.Header>
      <Accordion.Collapse className="settings-card__body" eventKey={form.id}>
        <Card.Body className="pt-2">
          <Form onSubmit={handleSubmit(updateSettingsSubmit)}>
            {!!form.note && (
              <p
                data-testid="form-note"
                dangerouslySetInnerHTML={{ __html: form.note }}
              />
            )}

            <div className="row mt-3">
              <div data-testid="form-container" className="col-12 col-sm-6">
                {form.fields.map(field => {
                  const visibilityOverrideField =
                    visibilityOverrideFields[field.id];
                  const visibilityOverride =
                    visibilityOverrideField &&
                    Object.keys(visibilityOverrideField).some(key =>
                      visibilityOverrideField[key].includes(
                        watchAllFields[key],
                      ),
                    );
                  if (
                    !(field.visible ?? true) &&
                    !(visibilityOverride ?? false)
                  ) {
                    return null;
                  }
                  if (field.id === 'submit_button') {
                    return (
                      <LoadingButton
                        key={field.id}
                        data-testid={field.id}
                        loading={!!formState.isSubmitting}
                        className="mt-2"
                        variant="primary"
                        type="submit"
                      >
                        {field.title}
                      </LoadingButton>
                    );
                  }
                  const isFieldSelect = field.type === 'select';
                  const formGroupAs = isFieldSelect ? 'select' : 'input';
                  const formGroupType = isFieldSelect ? 'text' : field.type;
                  const formGroupChildren =
                    field.type === 'select'
                      ? field.values?.map(option => (
                          <option key={option.id} value={option.id}>
                            {option.title}
                          </option>
                        ))
                      : null;
                  return (
                    <Form.Group key={field.id}>
                      <Form.Control
                        data-testid={field.id}
                        ref={register}
                        as={formGroupAs}
                        disabled={field.disabled}
                        size="sm"
                        type={formGroupType}
                        autoComplete={
                          field.type === 'password'
                            ? 'current-password'
                            : undefined
                        }
                        id={field.id}
                        name={field.id}
                        placeholder=" "
                      >
                        {formGroupChildren}
                      </Form.Control>
                      {!isFieldSelect && (
                        <>
                          <label
                            data-testid={`${field.id}-title`}
                            htmlFor="amount"
                          >
                            {field.title}
                          </label>
                          <div className="form-group__icons">
                            <i className="icon-check"></i>
                            <i className="icon-exclamation"></i>
                          </div>
                        </>
                      )}

                      {field.errors?.map(error => (
                        <small
                          data-testid={`${field.id}-error`}
                          className="form-group__error-msg"
                        >
                          {error}
                        </small>
                      ))}
                    </Form.Group>
                  );
                })}
              </div>
            </div>
          </Form>
        </Card.Body>
      </Accordion.Collapse>
      <i
        className="settings-card__icon icon-down1"
        onClick={accordionOnClick}
      />{' '}
    </Card>
  );
};

export default DynamicSettingsAccordion;
