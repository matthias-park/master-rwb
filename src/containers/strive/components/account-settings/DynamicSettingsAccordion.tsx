import React, { useCallback, useContext, useMemo } from 'react';
import { SettingsForm } from '../../../../types/api/user/ProfileSettings';
import { FormProvider, useForm } from 'react-hook-form';
import clsx from 'clsx';
import Accordion from 'react-bootstrap/Accordion';
import AccordionContext from 'react-bootstrap/AccordionContext';
import Form from 'react-bootstrap/Form';
import Card from 'react-bootstrap/Card';
import { useAccordionToggle } from 'react-bootstrap/AccordionToggle';
import LoadingButton from '../../../../components/LoadingButton';
import { useI18n } from '../../../../hooks/useI18n';
import { VALIDATIONS } from '../../../../constants';
import SelectInput from '../../../../components/customFormInputs/SelectInput';
import TextInput from '../../../../components/customFormInputs/TextInput';
interface SettingProps {
  form: SettingsForm;
  onSubmit: (
    url: string,
    body: { [key: string]: string | Blob },
    formBody?: boolean,
    updateUser?: boolean,
  ) => void;
}

const FormsWithUpdateUser = [
  'disable_player',
  'disable_player_time_out',
  'permanent_disable',
];

const DynamicSettingsAccordion = ({ form, onSubmit }: SettingProps) => {
  const { t } = useI18n();
  const currentEventKey = useContext(AccordionContext);
  const accordionOnClick = useAccordionToggle(form.id);
  const formMethods = useForm<any, any>({
    mode: 'onBlur',
  });
  const { handleSubmit, watch, formState } = formMethods;
  const watchAllFields = watch();
  const visibilityOverrideFields = useMemo(
    () =>
      form.fields
        .filter(
          field =>
            field.type === 'select' &&
            field.values?.some(value => value.additional_fields),
        )
        .reduce((obj, field) => {
          for (const value of field.values || []) {
            if (value.additional_fields) {
              for (const additionalField of value.additional_fields) {
                if (!obj[additionalField]) obj[additionalField] = {};
                if (!obj[additionalField][field.id])
                  obj[additionalField][field.id] = [];

                obj[additionalField][field.id].push(value.id.toString());
              }
            }
          }

          return obj;
        }, {}),
    [form.fields],
  );
  const updateSettingsSubmit = useCallback(
    data =>
      onSubmit(form.action, data, false, FormsWithUpdateUser.includes(form.id)),
    [],
  );

  return (
    <Card className="settings-card">
      <Card.Header
        className={clsx(
          'settings-card__header',
          currentEventKey === form.id && 'active',
        )}
      >
        <Accordion.Toggle
          data-testid="accordion-toggle"
          as="a"
          eventKey={form.id}
          className="text-dark"
        >
          {form.title}
        </Accordion.Toggle>
      </Card.Header>
      <Accordion.Collapse className="settings-card__body" eventKey={form.id}>
        <Card.Body className="pt-2">
          <FormProvider {...formMethods}>
            <Form onSubmit={handleSubmit(updateSettingsSubmit)}>
              {!!form.note && (
                <p
                  data-testid="form-note"
                  dangerouslySetInnerHTML={{ __html: form.note }}
                />
              )}

              <div className="row mt-3">
                <div data-testid="form-container" className="col-12 col-sm-6">
                  {form.fields.map(field => {
                    const visibilityOverrideField =
                      visibilityOverrideFields[field.id];
                    const visibilityOverride =
                      visibilityOverrideField &&
                      Object.keys(visibilityOverrideField).some(key =>
                        visibilityOverrideField[key].includes(
                          watchAllFields[key],
                        ),
                      );
                    if (
                      !(field.visible ?? true) &&
                      !(visibilityOverride ?? false)
                    ) {
                      return null;
                    }
                    switch (field.id) {
                      case 'submit_button': {
                        return (
                          <LoadingButton
                            key={field.id}
                            data-testid={field.id}
                            loading={!!formState.isSubmitting}
                            disabled={Object.values(watchAllFields).some(
                              value => !value || value === 'default',
                            )}
                            className="mt-2"
                            variant="primary"
                            type="submit"
                          >
                            {field.title}
                          </LoadingButton>
                        );
                      }
                      case 'select': {
                        return (
                          <SelectInput
                            key={field.id}
                            id={field.id}
                            rules={{
                              required: `${field.title} ${t(
                                'settings_field_required',
                              )}`,
                            }}
                            disabled={field.disabled}
                            defaultValue={field.default}
                            values={
                              field.values?.map(option => ({
                                value: option.id,
                                text: option.title,
                              })) || []
                            }
                            title={field.title}
                          />
                        );
                      }
                      default: {
                        const isPassword = [
                          'new_password',
                          'new_password_confirmation',
                        ].includes(field.id);
                        return (
                          <TextInput
                            key={field.id}
                            id={field.id}
                            rules={{
                              required: `${field.title} ${t(
                                'settings_field_required',
                              )}`,
                              validate: value => {
                                if (isPassword)
                                  return (
                                    VALIDATIONS.passwordMixOfThree(value) ||
                                    t('register_password_weak')
                                  );
                                if (field.id === 'phone_number')
                                  return (
                                    VALIDATIONS.phone(value) ||
                                    t('phone_number_invalid')
                                  );
                                return true;
                              },
                            }}
                            defaultValue={field.default}
                            disabled={field.disabled}
                            title={field.title}
                            toggleVisibility={isPassword}
                            type={field.type}
                          />
                        );
                      }
                    }
                  })}
                </div>
              </div>
            </Form>
          </FormProvider>
        </Card.Body>
      </Accordion.Collapse>
      <i
        className={clsx(
          `icon-${window.__config__.name}-down1`,
          'settings-card__icon',
        )}
        onClick={accordionOnClick}
      />{' '}
    </Card>
  );
};

export default DynamicSettingsAccordion;
