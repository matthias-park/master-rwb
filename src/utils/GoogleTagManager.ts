import {
  IDataLayer,
  ISnippets,
  ISendToGTM,
  ISetupGTM,
  ISnippetsParams,
} from '../types/GoogleTagManager';

const getDataLayerSnippet = (
  dataLayer: Pick<IDataLayer, 'dataLayer'>['dataLayer'],
  dataLayerName: Pick<
    IDataLayer,
    'dataLayerName'
  >['dataLayerName'] = 'dataLayer',
): Pick<ISnippets, 'gtmDataLayer'>['gtmDataLayer'] =>
  `window.${dataLayerName} = window.${dataLayerName} || []; window.${dataLayerName}.push(${JSON.stringify(
    dataLayer,
  )})`;

const getIframeSnippet = (
  id: Pick<ISnippetsParams, 'id'>['id'],
  environment?: Pick<ISnippetsParams, 'environment'>['environment'],
) => {
  let params = ``;
  if (environment) {
    const { gtm_auth, gtm_preview } = environment;
    params = `&gtm_auth=${gtm_auth}&gtm_preview=${gtm_preview}&gtm_cookies_win=x`;
  }
  return `<iframe src="https://www.googletagmanager.com/ns.html?id=${id}${params}" height="0" width="0" style="display:none;visibility:hidden" id="tag-manager"></iframe>`;
};

const getGTMScript = (
  dataLayerName: Pick<ISnippetsParams, 'dataLayerName'>['dataLayerName'],
  id: Pick<ISnippetsParams, 'id'>['id'],
  environment?: Pick<ISnippetsParams, 'environment'>['environment'],
) => {
  let params = ``;
  if (environment) {
    const { gtm_auth, gtm_preview } = environment;
    params = `+"&gtm_auth=${gtm_auth}&gtm_preview=${gtm_preview}&gtm_cookies_win=x"`;
  }
  return `
      (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
        new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
        j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
        'https://www.googletagmanager.com/gtm.js?id='+i+dl${params};f.parentNode.insertBefore(j,f);
      })(window,document,'script','${dataLayerName}','${id}');
    `;
};
/**
 * Function to setup the Google Tag Manager
 * @param params - The snippets params
 */
const setupGTM = (params: ISnippetsParams): ISetupGTM => {
  const getDataLayerScript = (): HTMLElement => {
    const dataLayerScript = document.createElement('script');
    dataLayerScript.innerHTML = getDataLayerSnippet(
      params.dataLayer,
      params.dataLayerName,
    );
    return dataLayerScript;
  };

  const getNoScript = (): HTMLElement => {
    const noScript = document.createElement('noscript');
    noScript.innerHTML = getIframeSnippet(params.id, params.environment);
    return noScript;
  };

  const getScript = (): HTMLElement => {
    const script = document.createElement('script');
    script.innerHTML = getGTMScript(
      params.dataLayerName,
      params.id,
      params.environment,
    );
    return script;
  };

  return {
    getDataLayerScript,
    getNoScript,
    getScript,
  };
};

/**
 * Function to init the GTM
 * @param dataLayer - The dataLayer
 * @param dataLayerName - The dataLayer name
 * @param environment - Specify the custom environment to use
 * @param id - The ID of the GTM
 */
export const initGTM = ({
  dataLayer,
  dataLayerName,
  environment,
  id,
}: ISnippetsParams): void => {
  const gtm = setupGTM({
    dataLayer,
    dataLayerName,
    environment,
    id,
  });

  const dataLayerScript = gtm.getDataLayerScript();
  const script = gtm.getScript();
  const noScript = gtm.getNoScript();

  document.head.insertBefore(script, document.head.childNodes[0]);
  document.head.insertBefore(dataLayerScript, document.head.childNodes[0]);
  document.body.insertBefore(noScript, document.body.childNodes[0]);
};

/**
 * Function to send the events to the GTM
 * @param dataLayerName - The dataLayer name
 * @param data - The data to push
 */
export const sendToGTM = ({ dataLayerName, data }: ISendToGTM): void =>
  window[dataLayerName].push(data);
