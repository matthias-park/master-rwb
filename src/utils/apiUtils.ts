import fetch from 'isomorphic-unfetch';
import { ConfigInterface, mutate } from 'swr';
import { RailsApiResponseFallback } from '../constants';
import RailsApiResponse from '../types/api/RailsApiResponse';
import { RegistrationPostalCodeAutofill } from '../types/api/user/Registration';
import { cleanPostBody } from '.';
import * as Sentry from '@sentry/react';

export const formatUrl = (
  url: string,
  params: { [key: string]: string | number | undefined } = {},
): string => {
  const paramsArr = Object.keys(params)
    .filter(key => params[key])
    .map(key => `${key}=${params[key]}`);
  return `${url}${paramsArr.length ? '?' : ''}${paramsArr.join('&')}`;
};

interface GetApiOptions {
  responseText?: boolean;
  cache?: RequestCache;
}

export const getApi = <T>(url: string, options?: GetApiOptions): Promise<T> => {
  const config: RequestInit = {
    mode: 'cors',
    credentials: 'include',
    cache: options?.cache,
    headers: new Headers(),
  };
  const getUrl = url.startsWith('http')
    ? url
    : `${window.__config__.apiUrl}${url}`;
  return fetch(getUrl, config).then(res => {
    if (!res.ok && res.status !== 400) {
      if (![401, 403].includes(res.status)) {
        Sentry.captureMessage(
          `Request failed ${url} with status ${res.status}`,
          Sentry.Severity.Fatal,
        );
      } else {
        mutate('/restapi/v1/user/status');
      }
      return Promise.reject<RailsApiResponse<null>>({
        ...RailsApiResponseFallback,
        Code: res.status,
      });
    }
    if (options?.responseText) {
      return res.text();
    }
    return res.json();
  });
};

export interface PostOptions {
  formData?: boolean;
}

export const postApi = <T>(
  url: string,
  body?: {
    [key: string]: number | string | Blob | boolean | undefined | null | object;
  },
  options: PostOptions = {},
): Promise<T> => {
  const config: RequestInit = {
    method: 'post',
    mode: 'cors',
    credentials: 'include',
    headers: new Headers(),
  };
  if (body) {
    if (options.formData) {
      const formData = new FormData();
      const formBody = cleanPostBody(body);
      for (const key in formBody) {
        if (Array.isArray(formBody[key])) {
          for (const item of formBody[key]) {
            formData.append(key, item as string | Blob);
          }
        } else {
          formData.append(key, formBody[key] as string | Blob);
        }
      }
      config.body = formData;
    } else {
      config.body = JSON.stringify(cleanPostBody(body));
      (config.headers as Headers).append('Content-Type', 'application/json');
    }
  }
  const postUrl = url.startsWith('http')
    ? url
    : `${window.__config__.apiUrl}${url}`;
  return fetch(postUrl, config).then(res => {
    if (!res.ok && res.status !== 400) {
      if (![401, 403].includes(res.status)) {
        Sentry.captureMessage(
          `Request failed ${url} with status ${res.status}`,
          Sentry.Severity.Fatal,
        );
      } else {
        mutate('/restapi/v1/user/status');
      }
      return Promise.reject<RailsApiResponse<null>>({
        ...RailsApiResponseFallback,
        Code: res.status,
      });
    }
    return res.json();
  });
};

export const SwrFetcherConfig: ConfigInterface<
  any,
  any,
  (...args: any) => any
> = {
  fetcher: getApi,
  revalidateOnFocus: false,
  revalidateOnReconnect: false,
  errorRetryCount: 2,
  dedupingInterval: 30000,
  focusThrottleInterval: 10000,
  errorRetryInterval: 10000,
  onError: (err, key) => {
    if (![401, 403].includes(err.status)) {
      Sentry.captureMessage(
        `Request failed ${key} with status ${err.status}`,
        Sentry.Severity.Fatal,
      );
    }
  },
};

export const formatSuccesfullRailsApiResponse = <T>(
  data: T,
): RailsApiResponse<T> => ({
  Code: 0,
  Data: data,
  Message: '',
  Success: true,
});

export const API_VALIDATIONS = {
  email: async (email: string): Promise<RailsApiResponse<{}>> => {
    const res = await postApi<RailsApiResponse<{}>>(
      '/restapi/v1/registration/check/email',
      {
        email,
      },
    ).catch(err => err);
    return res;
  },
  personalCode: async (
    personal_code: string,
  ): Promise<RailsApiResponse<{}>> => {
    const res = await postApi<RailsApiResponse<{}>>(
      '/restapi/v1/registration/check/personal_code',
      {
        personal_code,
      },
    ).catch((err: RailsApiResponse<{}>) => err);
    return res;
  },
  postalCode: async (
    post_code: string,
  ): Promise<RailsApiResponse<RegistrationPostalCodeAutofill | null>> => {
    const res = await postApi<
      RailsApiResponse<RegistrationPostalCodeAutofill | null>
    >('/restapi/v1/registration/check/post_code', {
      post_code,
    }).catch((err: RailsApiResponse<null>) => err);
    return res;
  },
};
