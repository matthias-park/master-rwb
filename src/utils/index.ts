import { PagesName } from '../constants';
import { PageConfig } from '../types/api/PageConfig';
import Lockr from 'lockr';
import { PostItem } from '../types/api/Posts';
import dayjs from 'dayjs';

export const sortAscending = (a: number, b: number) => a - b;
export const sortDescending = (a: number, b: number) => b - a;

export function throttle(func: Function, limit: number): Function {
  let inThrottle: boolean;

  return function (this: any): any {
    const args = arguments;
    const context = this;
    if (!inThrottle) {
      inThrottle = true;
      func.apply(context, args);
      setTimeout(() => (inThrottle = false), limit);
    }
  };
}

export const stringToMiliseconds = (time: string): number | null => {
  const parsed = time.split(':');
  if (parsed.length === 3) {
    return miliseconds(Number(parsed[0]), Number(parsed[1]), Number(parsed[2]));
  }
  return null;
};
export const miliseconds = (
  hrs: number = 0,
  min: number = 0,
  sec: number = 0,
): number => {
  return (hrs * 60 * 60 + min * 60 + sec) * 1000;
};
export const removeFalsyFromObject = (obj: any) =>
  Object.keys(obj).forEach(
    k => !obj[k] && obj[k] !== undefined && delete obj[k],
  );

export const structuredBankCommunications = (barcode): string => {
  if (!barcode) return '';
  const rExp = /\d+/;
  const moduloOf97 = barcode.match(rExp)[0] % 97;
  return '00' + barcode + moduloOf97.toString().padStart(2, '0');
};

export const formatNavigationRoutes = (constants: PageConfig) => {
  const navigationRoutes = constants.navigation_routes
    ? [...constants.navigation_routes]
    : [];
  if (constants.content_pages) {
    for (const [key, value] of Object.entries(constants.content_pages)) {
      const name = Array.isArray(value) ? value[0] : value;
      const slug = Array.isArray(value) ? value[1] : key;
      const contentPagePath = `${slug.startsWith('/') ? '' : '/'}${slug}`;
      if (!navigationRoutes.some(route => route.path === contentPagePath)) {
        navigationRoutes.push({
          id:
            Array.isArray(value) && PagesName[key]
              ? Number(key)
              : PagesName.TemplatePage,
          path: contentPagePath,
          name,
        });
      }
    }
  }
  return navigationRoutes.sort((a, b) =>
    sortDescending(a.path.length, b.path.length),
  );
};

export const clearUserLocalStorage = () => {
  const keys = ['transactions-date-to', 'transactions-date-from'];
  for (const key of keys) {
    localStorage.removeItem(key);
  }
};

export const cleanPostBody = (obj: { [key: string]: unknown }) =>
  Object.entries(obj).reduce((obj, [key, value]) => {
    if (!key.includes('temp_field_') && value != null) {
      obj[key] = typeof value === 'string' ? value.trim() : value;
    }
    return obj;
  }, {});

export const errorHandler = (event: ErrorEvent) => {
  if (event.error?.stack?.includes('kambi-chunk')) {
    const kambiErrorId = 'kambi-error-reload';
    const kambiErrorRetryCount = Lockr.get(kambiErrorId, 0) + 1;
    if (kambiErrorRetryCount < 3) {
      Lockr.set(kambiErrorId, kambiErrorRetryCount);
      window.location.reload();
    }
  }
};

export const filterPromotionsList = (promotions: PostItem[]): PostItem[] =>
  promotions.filter(promotion => {
    const isNotExpired =
      !promotion.expiration_date ||
      dayjs(promotion.expiration_date).isAfter(dayjs());
    const isPublished =
      promotion.publish_date && dayjs(promotion.publish_date).isBefore(dayjs());
    const isVisible = promotion.visible;
    return isNotExpired && isPublished && isVisible;
  });
