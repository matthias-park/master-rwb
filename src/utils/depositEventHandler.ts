import { CustomWindowEvents, ComponentName } from '../constants';
import RequestReturn from '../types/api/deposits/RequestReturn';
import RailsApiResponse from '../types/api/RailsApiResponse';
import depositEventHandlerProps from '../types/DepositEventHandler';
import { postApi } from './apiUtils';
import * as Sentry from '@sentry/react';

const depositEventHandler = ({
  eventType,
  eventData,
  setCustomHtml,
  depositStatus,
  setApiError,
  setDepositLoading,
  t,
  depositRequestId,
  bankId,
  locale,
  setCurrentStep,
  disableModal,
  depositForm,
}: depositEventHandlerProps) => {
  switch (eventType) {
    case CustomWindowEvents.DepositPaymentSuccess: {
      setApiError(null);
      setCustomHtml(null);
      depositStatus.startCheckingStatus();
      break;
    }
    case CustomWindowEvents.DepositPaymentError: {
      if (typeof eventData === 'string') {
        Sentry.captureMessage(`deposit error: ${eventData}`);
        setApiError({ message: eventData, variant: 'danger' });
      } else {
        setApiError({ message: t('api_response_failed'), variant: 'danger' });
      }
      setCustomHtml(null);
      setDepositLoading(false);
      break;
    }
    case CustomWindowEvents.DepositPaymentWarning: {
      if (typeof eventData === 'string') {
        setApiError({ message: eventData, variant: 'warning' });
      }
      break;
    }
    case CustomWindowEvents.DepositRequestReturn: {
      if (eventData && typeof eventData === 'object') {
        const data: any = {
          ...eventData,
          depositRequestId,
        };
        if (bankId === 189) {
          data.browserInfo = {
            acceptHeader: '*',
            colorDepth: window.screen.colorDepth,
            javaEnabled: false,
            language: locale,
            screenHeight: window.innerHeight,
            screenWidth: window.innerWidth,
            timeZoneOffset: new Date().getTimezoneOffset(),
            userAgent: window.navigator.userAgent,
          };
        }
        postApi<RailsApiResponse<RequestReturn>>(
          '/restapi/v1/deposits/request_return',
          {
            bank_id: bankId.toString(),
            data: JSON.stringify(data),
          },
        )
          .then(res => {
            if (res.Data.OK) {
              if (res.Data.AdditionalData?.redirectUrl) {
                if (depositForm) {
                  window.open(res.Data.AdditionalData.redirectUrl, '_blank');
                  disableModal?.(ComponentName.QuickDepositModal);
                } else {
                  window.location.href = res.Data.AdditionalData.redirectUrl;
                }
                return;
              }
              if (res.Data.AdditionalData?.verifyPaymentData) {
                return window.postMessage(
                  JSON.stringify({
                    eventType: CustomWindowEvents.DepositVerifyPayment,
                    eventData: res.Data.AdditionalData.verifyPaymentData,
                  }),
                  '*',
                );
              }
              setCustomHtml(null);
              depositStatus.startCheckingStatus();
            } else if (res.Data.Error) {
              setCurrentStep?.(1);
              setApiError({
                message: t('api_response_failed'),
                variant: 'danger',
              });
              Sentry.captureMessage(
                `Got DepositRequestReturn error: ${res.Data.Error}`,
                Sentry.Severity.Info,
              );
            }
          })
          .catch(() => null);
      } else {
        Sentry.captureMessage(
          'Got DepositRequestReturn without data',
          Sentry.Severity.Fatal,
        );
      }
    }
  }
};
export default depositEventHandler;
