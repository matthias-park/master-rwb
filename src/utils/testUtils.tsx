import React, { useEffect, useState } from 'react';
import { render } from '@testing-library/react';
import { SWRConfig } from 'swr';
import { SwrFetcherConfig } from './apiUtils';
import { ToastProvider } from 'react-toast-notifications';
import { HelmetProvider } from 'react-helmet-async';
import { Provider } from 'react-redux';
import createStoreAsync from '../state';

const AllTheProviders = ({ children }) => {
  const [state, setState] = useState<any>(null);
  useEffect(() => {
    createStoreAsync().then(store => setState(store));
  }, []);
  if (!state) return null;
  return (
    <Provider store={state}>
      <SWRConfig value={SwrFetcherConfig}>
        <ToastProvider>
          <HelmetProvider>{children}</HelmetProvider>
        </ToastProvider>
      </SWRConfig>
    </Provider>
  );
};

const customRender = (ui, options: any = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
