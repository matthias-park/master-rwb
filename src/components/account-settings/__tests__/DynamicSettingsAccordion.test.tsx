import React from 'react';
import { render, cleanup, fireEvent } from '../../../utils/testUtils';
import renderer from 'react-test-renderer';
import DynamicSettingsAccordion from '../DynamicSettingsAccordion';
import { act } from 'react-dom/test-utils';
import {
  SettingsFieldStyle,
  SettingsForm,
} from '../../../types/api/user/ProfileSettings';

const form: SettingsForm = {
  id: 'max_bet_per_period_amount',
  title: 'Set the wagering amount limit per period',
  note:
    'This option allows you to set the limit on the amount you can wager per selected time period. Please note that after setting this limit you will not be able to wager more than the specified amount. Should you wish to reduce this amount, the changes will take effect immediately. However, if you want to increase or cancel the limit, the changes will only take effect after one week.',
  action:
    'https://bnl-dev.tglab.dev/v2/profile/set-max-bet-per-period-amount.json',
  errors: [],
  fields: [
    {
      id: 'max_bet_per_period_option',
      title: 'Setting',
      errors: [],
      type: 'select',
      default: 'default',
      values: [
        {
          title: 'Turn on/off',
          id: 'default',
        },
        {
          title: 'Turn on',
          id: 1,
          additional_fields: [
            'max_bet_per_period_period',
            'max_bet_per_period_amount',
          ],
        },
        {
          title: 'Turn off',
          id: 2,
          additional_fields: ['max_bet_per_period_period'],
        },
      ],
    },
    {
      id: 'max_bet_per_period_period',
      title: 'Choose period:',
      errors: [],
      type: 'select',
      default: 'default',
      visible: false,
      values: [
        {
          title: 'Choose period',
          id: 'default',
        },
        {
          title: '1 week',
          id: 1,
        },
      ],
    },
    {
      id: 'max_bet_per_period_amount',
      title: 'Amount',
      visible: false,
      errors: [],
      type: 'text',
    },
    {
      id: 'max_bet_per_period_password',
      title: 'Password',
      errors: [],
      type: 'password',
    },
    {
      id: 'submit_button',
      style: SettingsFieldStyle.Positive,
      title: 'Set',
      type: 'submit',
    },
  ],
};

beforeEach(cleanup);

test('displays title correctly', async () => {
  const { getByTestId } = render(
    <DynamicSettingsAccordion form={form} onSubmit={() => {}} />,
  );
  expect(getByTestId('accordion-toggle')).toHaveTextContent(form.title);
});

test('displays note correctly', async () => {
  const { getByTestId } = render(
    <DynamicSettingsAccordion form={form} onSubmit={() => {}} />,
  );
  expect(getByTestId('form-note')).toHaveTextContent(form.note);
});

test('contains 3 inputs by default', async () => {
  const { getByTestId } = render(
    <DynamicSettingsAccordion form={form} onSubmit={() => {}} />,
  );
  expect(getByTestId('form-container').children).toHaveLength(3);
});

test('additional fields are displayed', async () => {
  const { getByTestId } = render(
    <DynamicSettingsAccordion form={form} onSubmit={() => {}} />,
  );
  fireEvent.change(getByTestId('max_bet_per_period_option'), {
    target: { value: 1 },
  });
  expect(getByTestId('form-container').children).toHaveLength(5);
});

test('submit value correctly', async () => {
  const data = {
    max_bet_per_period_option: '1',
    max_bet_per_period_period: '1',
    max_bet_per_period_amount: '123',
    max_bet_per_period_password: 'secret_password',
  };
  const { getByTestId } = render(
    <DynamicSettingsAccordion
      form={form}
      onSubmit={(url, body) => {
        expect(url).toBe(form.action);
        expect(body).toStrictEqual(data);
      }}
    />,
  );
  for (const key in data) {
    fireEvent.change(getByTestId(key), {
      target: { value: data[key] },
    });
  }
  await act(async () => {
    fireEvent.click(getByTestId('submit_button'));
  });
});

test('default matches snapshot', () => {
  const dom = renderer.create(
    <DynamicSettingsAccordion form={form} onSubmit={() => {}} />,
  );
  expect(dom).toMatchSnapshot();
});
