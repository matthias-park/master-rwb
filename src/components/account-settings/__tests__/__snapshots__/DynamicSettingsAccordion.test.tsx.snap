// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default matches snapshot 1`] = `
<div
  className="settings-card card"
>
  <div
    className="settings-card__header card-header"
  >
    <a
      className="text-dark"
      data-testid="accordion-toggle"
      onClick={[Function]}
    >
      Set the wagering amount limit per period
    </a>
  </div>
  <div
    aria-expanded={null}
    className="settings-card__body collapse"
  >
    <div
      className="pt-2 card-body"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "This option allows you to set the limit on the amount you can wager per selected time period. Please note that after setting this limit you will not be able to wager more than the specified amount. Should you wish to reduce this amount, the changes will take effect immediately. However, if you want to increase or cancel the limit, the changes will only take effect after one week.",
            }
          }
          data-testid="form-note"
        />
        <div
          className="row mt-3"
        >
          <div
            className="col-12 col-sm-6"
            data-testid="form-container"
          >
            <div
              className="form-group"
            >
              <select
                className="form-control form-control-sm"
                data-testid="max_bet_per_period_option"
                id="max_bet_per_period_option"
                name="max_bet_per_period_option"
                placeholder=" "
                type="text"
              >
                <option
                  value="default"
                >
                  Turn on/off
                </option>
                <option
                  value={1}
                >
                  Turn on
                </option>
                <option
                  value={2}
                >
                  Turn off
                </option>
              </select>
            </div>
            <div
              className="form-group"
            >
              <input
                autoComplete="current-password"
                className="form-control form-control-sm"
                data-testid="max_bet_per_period_password"
                id="max_bet_per_period_password"
                name="max_bet_per_period_password"
                placeholder=" "
                type="password"
              />
              <label
                data-testid="max_bet_per_period_password-title"
                htmlFor="amount"
              >
                Password
              </label>
              <div
                className="form-group__icons"
              >
                <i
                  className="icon-check"
                />
                <i
                  className="icon-exclamation"
                />
              </div>
            </div>
            <button
              className="mt-2 btn btn-primary"
              data-testid="submit_button"
              disabled={false}
              type="submit"
            >
              Set
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
  <i
    className="settings-card__icon icon-down1"
    onClick={[Function]}
  />
   
</div>
`;
